kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CubeBuildJob.java

```diff
@@ -319,45 +324,61 @@ private void build(Collection<NBuildSourceInfo> buildSourceInfos, SegmentInfo se
     // build current layer and return the next layer to be built.
     private List<NBuildSourceInfo> buildLayer(Collection<NBuildSourceInfo> buildSourceInfos, SegmentInfo seg,
                                               SpanningTree st) {
-        int cuboidsNumInLayer = 0;
 
-        // build current layer
+        int cuboidsNumInLayer = 0;
         List<LayoutEntity> allIndexesInCurrentLayer = new ArrayList<>();
+
+        //record build infos before building
         for (NBuildSourceInfo info : buildSourceInfos) {
             Collection<LayoutEntity> toBuildCuboids = info.getToBuildCuboids();
             infos.recordParent2Children(info.getLayout(),
                     toBuildCuboids.stream().map(LayoutEntity::getId).collect(Collectors.toList()));
             cuboidsNumInLayer += toBuildCuboids.size();
             Preconditions.checkState(!toBuildCuboids.isEmpty(), "To be built cuboids is empty.");
-            Dataset<Row> parentDS = info.getParentDS();
-            // record the source count of flat table
-            if (info.getLayoutId() == ParentSourceChooser.FLAT_TABLE_FLAG()) {
-                cuboidsRowCount.putIfAbsent(info.getLayoutId(), parentDS.count());
-            }
+            info.getToBuildCuboids().stream().forEach(allIndexesInCurrentLayer::add);
+            infos.recordCuboidsNumPerLayer(seg.id(), cuboidsNumInLayer);
```

Comment:
the value of 'cuboidsNumInLayer' is the total count of current segments, why needs to set this value into infos per loop